internal/controller/pipelinerollout_controller.go

```diff
@@ -236,3 +302,125 @@ func (r *PipelineRolloutReconciler) SetupWithManager(mgr ctrl.Manager) error {
 
 	return nil
 }
+
+func setPipelineDesiredStatus(obj *kubernetes.GenericObject, status string) (*kubernetes.GenericObject, error) {
+	unstruc, err := kubernetes.ObjectToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	err = unstructured.SetNestedField(unstruc.Object, status, "spec", "lifecycle", "desiredPhase")
+	if err != nil {
+		return nil, err
+	}
+
+	newObj, err := kubernetes.UnstructuredToObject(unstruc)
+	if err != nil {
+		return nil, err
+	}
+	return newObj, nil
+}
+
+func isPipelinePausing(ctx context.Context, pipeline *kubernetes.GenericObject) bool {
+	numaLogger := logger.FromContext(ctx)
+	pipelineStatus, err := kubernetes.ParseStatus(pipeline)
+	if err != nil {
+		numaLogger.Errorf(err, "failed to parse Pipeline Status from pipeline CR: %+v, %v", pipeline, err)
+		return false
+	}
+
+	numaLogger.Debugf("pipeline status: %+v", pipelineStatus)
+
+	return numaflowv1.PipelinePhase(pipelineStatus.Phase) == numaflowv1.PipelinePhasePausing
+}
+
+func isPipelinePaused(ctx context.Context, pipeline *kubernetes.GenericObject) bool {
+	numaLogger := logger.FromContext(ctx)
+	pipelineStatus, err := kubernetes.ParseStatus(pipeline)
+	if err != nil {
+		numaLogger.Errorf(err, "failed to parse Pipeline Status from pipeline CR: %+v, %v", pipeline, err)
+		return false
+	}
+
+	numaLogger.Debugf("pipeline status: %+v", pipelineStatus)
+
+	return numaflowv1.PipelinePhase(pipelineStatus.Phase) == numaflowv1.PipelinePhasePaused
+}
+
+func applyPipelineSpec(
+	ctx context.Context,
+	restConfig *rest.Config,
+	obj *kubernetes.GenericObject,
+	pipelineRollout *apiv1.PipelineRollout,
+) error {
+	numaLogger := logger.FromContext(ctx)
+	// TODO: use UpdateSpec instead
+	err := kubernetes.ApplyCRSpec(ctx, restConfig, obj, "pipelines")
+	if err != nil {
+		numaLogger.Errorf(err, "failed to apply Pipeline: %v", err)
+		pipelineRollout.Status.MarkFailed("ApplyPipelineFailure", err.Error())
+		return err
+	}
+	// after the Apply, Get the Pipeline so that we can propagate its health into our Status
+	pipeline, err := kubernetes.GetCR(ctx, restConfig, obj, "pipelines")
+	if err != nil {
+		numaLogger.Errorf(err, "failed to get Pipeline: %v", err)
+		return err
+	}
+
+	processPipelineStatus(ctx, pipeline, pipelineRollout)
+	return nil
+}
+
+// TODO: detect engine
+func needsPausing(obj *kubernetes.GenericObject, newObj *kubernetes.GenericObject) (bool, error) {
```

Comment:
since this is an optimization, I wonder if when we merge this in, we should just have this return true for now? (like first make sure we can always just pause reliably before trying to get fancy)

Reply:
yes :)